spark/src/main/java/org/apache/iceberg/actions/PlanScanAction.java

```diff
@@ -0,0 +1,263 @@
+    } else {
+      splitSize = meta.propertyAsLong(TableProperties.SPLIT_SIZE, TableProperties.SPLIT_SIZE_DEFAULT);
+    }
+    int lookback;
+    if (options.containsKey(TableProperties.SPLIT_LOOKBACK)) {
+      lookback = Integer.parseInt(options.get(TableProperties.SPLIT_LOOKBACK));
+    } else {
+      lookback = meta.propertyAsInt(TableProperties.SPLIT_LOOKBACK, TableProperties.SPLIT_LOOKBACK_DEFAULT);
+    }
+    long openFileCost;
+    if (options.containsKey(TableProperties.SPLIT_OPEN_FILE_COST)) {
+      openFileCost = Long.parseLong(options.get(TableProperties.SPLIT_OPEN_FILE_COST));
+    } else {
+      openFileCost = meta.propertyAsLong(
+          TableProperties.SPLIT_OPEN_FILE_COST, TableProperties.SPLIT_OPEN_FILE_COST_DEFAULT);
+    }
+
+    CloseableIterable<FileScanTask> fileScanTasks = planFiles();
+    CloseableIterable<FileScanTask> splitFiles = TableScanUtil.splitFiles(fileScanTasks, splitSize);
+    return TableScanUtil.planTasks(splitFiles, splitSize, lookback, openFileCost);
+  }
+
+  private CloseableIterable<FileScanTask> planFiles() {
+    if (snapshot != null) {
+      LOG.info("Using distributed planning for scanning table {} snapshot {} created at {} with filter {}",
+          scan.table(), snapshot.snapshotId(), snapshot.timestampMillis(), scan.filter());
+      Listeners.notifyAll(
+          new ScanEvent(scan.table().toString(), snapshot.snapshotId(), scan.filter(), scan.schema()));
+      return planFilesFor();
+    } else {
+      LOG.info("Scanning empty table {}", scan.table());
+      return CloseableIterable.empty();
+    }
+  }
+
+  private CloseableIterable<FileScanTask> planFilesFor() {
+    // This is different for different BaseTableScan implementations
+    if (scan instanceof DataTableScan) {
+      return planDataTableScanFiles();
+    } else {
+      throw new UnsupportedOperationException(String.format("Unable to do distributed planning for %s",
+          scan.getClass()));
+    }
+  }
+
+  private DeleteFileIndex buildDeleteFileIndex() {
+    // Build delete index locally
+    List<ManifestFile> deleteManifests = snapshot.deleteManifests();
+    DeleteFileIndex.Builder deleteFileIndexBuilder = DeleteFileIndex.builderFor(table.io(), deleteManifests);
+    deleteFileIndexBuilder.caseSensitive(scan.isCaseSensitive());
+    deleteFileIndexBuilder.specsById(table.specs());
+    deleteFileIndexBuilder.filterData(scan.filter());
+    return deleteFileIndexBuilder.build();
+  }
+
+  private CloseableIterable<FileScanTask> planDataTableScanFiles() {
+    // TODO Currently this approach reads all manifests, no manifest filtering - Maybe through pushdowns or table read
+    // options?
+
+
+    // Create a dataframe of all DataFile entries
+    String dataFilesMetadataTable = metadataTableName(MetadataTableType.ENTRIES);
+    Dataset<Row> manifestEntries =
+        spark.read()
+            .format("iceberg")
+            .load(dataFilesMetadataTable);
```

Comment:
I think this needs to set `snapshot-id` or `as-of-timestamp` if either one was set in the scan. Otherwise, this just uses the latest table state.

Reply:
If so we need to add some tests, Currently this passes pretty much everything in Spark2 Suit and Spark 3 suite with the exception of a few tests which make sure only a single scan is done. I'll double check this when I add in a DistributedPlanning Specific test suite, rather than just trying it against all existing tests.